extensions/amp-viewer-integration/0.1/amp-viewer-integration.js

```diff
@@ -44,54 +45,109 @@ export class AmpViewerIntegration {
 
     /** @private {?string|undefined} */
     this.unconfirmedViewerOrigin_ = null;
+
+    /** @private {boolean} */
+    this.isWebView_;
   }
 
   /**
    * Initiate the handshake. If handshake confirmed, start listening for
    * messages. The service is disabled if the viewerorigin parameter is
    * absent.
-   * @return {?Promise}
+   * @return {!Promise<undefined>}
    */
   init() {
     dev().fine(TAG, 'handshake init()');
     const viewer = viewerForDoc(this.win.document);
+    this.isWebView_ = viewer.getParam('webview') == 1;
+    this.isWebView_ = true; //delete this before submitting
     this.unconfirmedViewerOrigin_ = viewer.getParam('origin');
-    if (!this.unconfirmedViewerOrigin_) {
-      dev().fine(TAG, 'Viewer origin not specified.');
-      return null;
+
+    const win = this.win;
+    const unconfirmedViewerOrigin = this.unconfirmedViewerOrigin_;
+    if (!this.isWebView_) {
+      const port = new WindowPortEmulator();
+      port.addEventListener = function(eventType, handler) {
+        listen(win, 'message', e => {
+          if (e.origin == unconfirmedViewerOrigin &&
+              e.source == win.parent && e.data.app == APP) {
+            handler(e);
+          }
+        });
+      };
+      port.postMessage = function(data) {
+        win.parent./*OK*/postMessage(data, unconfirmedViewerOrigin);
+      };
+      return this.openChannelAndStart_(viewer, port);
     }
 
-    dev().fine(TAG, 'listening for messages', this.unconfirmedViewerOrigin_);
-    const messaging = new Messaging(
-      this.win, this.win.parent, this.unconfirmedViewerOrigin_);
+    // port/webview case
+    return this.webviewPreHandshakePromise_().then(receivedPort => {
+      const port = new WindowPortEmulator();
+      port.addEventListener = function(eventType, handler) {
+        receivedPort.onmessage = handler;
+      };
+      port.postMessage = function(data) {
+        receivedPort./*OK*/postMessage(data);
+      };
+      return this.openChannelAndStart_(viewer, port);
+    });
+  }
+
+  /**
+   * @return {!Promise}
+   * @private
+   */
+  webviewPreHandshakePromise_() {
```

Comment:
Add `origin` and `source` as args. For a WebView the values will be empty string and `null` resp.